Fix: The table cell should always have paragraph in the model.

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: The table cell should always have paragraph in the model. Closes #125.

BREAKING CHANGE: The `injectTablePostFixer()` function from  `table/converters/table-post-fixer` is now `injectTableLayoutPostFixer()`and is moved to `table/converters/table-layout-post-fixer` module.

---

### Additional information

* I've added you @Reinmar for the review - I'm not sure if the fix is sufficient.

Reinmar commented on src/converters/table-post-fixer.js:
```diff
@@ -195,23 +196,41 @@ import TableWalker from './../tablewalker';
  *			</tbody>
  *		</table>
  *
+ * ## Ensuring proper table structure
```
table structure? or rather "Ensuring a proper content of table cells"?

Reinmar commented on src/converters/table-post-fixer.js:
```diff
@@ -49,14 +50,14 @@ import TableWalker from './../tablewalker';
  *				<tr>
- *					<td rowspan="2">FOO</td>
- *					<td colspan="2">BAR</td>
+ *					<td rowspan="2"><p>FOO</p></td>
```
It seems that my comment disappeared so one more time...

Is it correct? Those `<p>`s won't be rendered (there will be spans or nothing). It's a bit confusing and I'm not sure if this is a mistake or a mental shortcut.

jodator replied:
I'd remove them.

Reinmar commented on tests/converters/table-post-fixer.js:
```diff
@@ -181,6 +181,19 @@ describe( 'Table post-fixer', () => {
 	} );
 
+	describe( 'on removing tableCell contents', () => {
+		it( 'should fix empty table cells', () => {
```
It's hard to understand from this test that it checks that a paragraph should be added. I checked that it fails without the patch and the error is quite obvious then, but the code and description of this test are cryptic. 

There are many opinions regarding writing so many tools for creating tests. But, leaving this aside, when we write them we should make sure that tests remain understandable.

Reinmar commented:
Other than the issues in the code, I checked that this PR fixes the bug.

jodator commented:
@Reinmar I've updated the solution:
- the Table Post-Ffixer was renamed to a more descriptive: Table Layout Post-Fixer as it fixes table layout (missing table cells)
- I've moved the table cell content post fixer to own file as it does something else then table layout post-fixer
- more tests added as it should act also on adding table/tableRow/tableCell to the model.

Reinmar commented on src/converters/table-cell-content-post-fixer.js:
```diff
@@ -0,0 +1,112 @@
+// @param {module:engine/model/element~Element} table
+// @param {module:engine/model/writer~Writer} writer
+function fixTableCellContent( tableCell, writer ) {
+	if ( tableCell.childCount == 0 ) {
```
What if a cell already has some content, but that's a text? We can make this a followup because I don't know right now scenario which would lead to this, but it's certainly possible.

jodator replied:
AFAIR the `$text` is not allowed in the model of a table cell.

Reinmar replied:
Which does not mean it cannot end there. Schema rules are not enforced. Check out https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/framework/guides/deep-dive/schema.html#implementing-additional-constraints and the following section.
